crates/net/src/builder.rs

```diff
@@ -193,6 +202,14 @@ impl NetworkDriverBuilder {
     }
 }
 
+fn listen_config_to_socket_addr(listen_config: &ListenConfig) -> Result<SocketAddr> {
+    match listen_config {
+        ListenConfig::Ipv4 { ip, port } => Ok(SocketAddr::new(IpAddr::V4(*ip), *port)),
+        ListenConfig::Ipv6 { ip, port } => Ok(SocketAddr::new(IpAddr::V6(*ip), *port)),
+        _ => Err(eyre::eyre!("invalid listen config")),
+    }
+}
```

Comment:
I don't think we need this - I think right now we should make it clear in the top level builder that the socket address is always used for the gossip driver and the listen config is used for discovery, but will fallback to using the same socket addr as the gossip driver.

Maybe we change the builder methods to be
- `with_gossip_addr`
- `with_discovery_addr`